File: aspnet-in-practice/06/CS/CustomControls.Templated/Message.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace CustomControls.Templated
{
	[ParseChildren(true)]
	public class Message : Control, INamingContainer
	{
		[TemplateContainer(typeof(MessageItem)), PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(typeof(ITemplate), "")]
		public ITemplate ItemTemplate { get; set; }

		public string Text
		{
			get { return ViewState["Text"] as string; }
			set { ViewState["Text"] = value; }
		}

		protected override void CreateChildControls()
		{
			if (ItemTemplate != null)
			{
				MessageItem template = new MessageItem(this);
				ItemTemplate.InstantiateIn(template);
				this.Controls.Clear();
				this.Controls.Add(template);
			}
			else
			{
				Controls.Add(new LiteralControl(Text));
			}
		}

		protected override void OnDataBinding(EventArgs e)
		{
			EnsureChildControls();
			base.OnDataBinding(e);
		}
	}
}

File: aspnet-in-practice/06/VB/CustomControls.Web/CustomControls.Templated/Message.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Web
Imports System.Web.UI
Imports System.Web.UI.WebControls

<ParseChildren(True)>
Public Class Message
	Inherits Control
	Implements INamingContainer

	<TemplateContainer(GetType(MessageItem)), PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(GetType(ITemplate), "")> _
	Public Property ItemTemplate() As ITemplate

	Public Property Text() As String
		Get
			Return TryCast(ViewState("Text"), String)
		End Get
		Set(ByVal value As String)
			ViewState("Text") = value
		End Set
	End Property

	Protected Overrides Sub CreateChildControls()
		If ItemTemplate IsNot Nothing Then
			Dim template As New MessageItem(Me)
			ItemTemplate.InstantiateIn(template)
			Me.Controls.Clear()
			Me.Controls.Add(template)
		Else
			Controls.Add(New LiteralControl(Text))
		End If
	End Sub

	Protected Overrides Sub OnDataBinding(ByVal e As EventArgs)
		EnsureChildControls()
		MyBase.OnDataBinding(e)
	End Sub
End Class